<h1 style="text-align: center;">Supermarket Customers</h1>
<h3 style="text-align: center;">Jawed Iqbal Alfaruqiy</h3>

---

# Latar Belakang
Supermarket memiliki data pelanggan yang mencakup demografi, pola pembelian, dan respons kampanye pemasaran. Memahami perilaku pelanggan penting untuk meningkatkan strategi pemasaran dan penjualan. Dengan analisis data, supermarket dapat mengoptimalkan pemasaran, meningkatkan retensi pelanggan, dan mengelola stok lebih efisien melalui segmentasi pelanggan dan promosi yang dipersonalisasi.

## Pernyataan Masalah

Supermarket ingin mengetahui **segmen pelanggan yang paling bernilai dan pola pembelian mereka**. Informasi ini akan membantu dalam meningkatkan efektivitas kampanye pemasaran, mengurangi biaya promosi yang tidak tepat sasaran, dan mengoptimalkan pengelolaan stok barang.

Sebagai seorang *data analyst*, kita akan mencoba menjawab pertanyaan berikut:

**Bagaimana karakteristik pelanggan yang memberikan kontribusi terbesar terhadap pendapatan supermarket, terutama dibandingkan pelanggan dengan pengeluaran lebih rendah?**

## Key Objective

Key Objective:

Untuk mengatasi permasalahan tersebut, beberapa tujuan utama dari analisis data ini adalah:

1. Mengidentifikasi segmen pelanggan berdasarkan data demografi dan pola pembelian guna meningkatkan personalisasi pemasaran.

1. Menganalisis efektivitas kampanye pemasaran dengan mengevaluasi tingkat respons pelanggan terhadap berbagai promosi.

1. Mengungkap pola pembelian pelanggan untuk membantu pengambilan keputusan dalam penawaran produk dan strategi harga.

1. Memprediksi loyalitas pelanggan dan mengembangkan strategi untuk meningkatkan keterlibatan serta retensi pelanggan.

# Data
Untuk menjawab pertanyaan di atas, kita akan menganalisa data peserta yang sudah dikumpulkan oleh perusahaan. Dataset dapat diakses [Disini](https://drive.google.com/drive/folders/1WodnBbuYTvsF0-6HTuQABQ0KCS31lqbK)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from scipy.stats import normaltest, chi2_contingency, mannwhitneyu, ttest_ind
from datetime import datetime

import warnings
warnings.filterwarnings("ignore")

Supermarket Customers Data Dicitonary:  
- Customer
    - ID: Customer's unique identifier
    - Year_Birth: Customer's birth year
    - Education: Customer's education level
    - Marital_Status: Customer's marital status
    - Income: Customer's yearly household income
    - Kidhome: Number of children in customer's household
    - Teenhome: Number of teenagers in customer's household
    - Dt_Customer: Date of customer's enrollment with the company
    - Recency: Number of days since customer's last purchase
    - Complain: 1 if the customer complained in the last 2 years, 0 otherwise
- Products
    - MntWines: Amount spent on wine in last 2 years
    - MntFruits: Amount spent on fruits in last 2 years
    - MntMeatProducts: Amount spent on meat in last 2 years
    - MntFishProducts: Amount spent on fish in last 2 years
    - MntSweetProducts: Amount spent on sweets in last 2 years
    - MntGoldProds: Amount spent on gold in last 2 years
- Promotion
    - NumDealsPurchases: Number of purchases made with a discount
    - AcceptedCmp1: 1 if the customer accepted the offer in the 1st campaign, 0 otherwise
    - AcceptedCmp2: 1 if the customer accepted the offer in the 2nd campaign, 0 otherwise
    - AcceptedCmp3: 1 if the customer accepted the offer in the 3rd campaign, 0 otherwise
    - AcceptedCmp4: 1 if the customer accepted the offer in the 4th campaign, 0 otherwise
    - AcceptedCmp5: 1 if the customer accepted the offer in the 5th campaign, 0 otherwise
    - Response: 1 if the customer accepted the offer in the last campaign, 0 otherwise
- Place
    - NumWebPurchases: Number of purchases made through the company’s website
    - NumCatalogPurchases: Number of purchases made using a catalog
    - NumStorePurchases: Number of purchases made directly in stores
    - NumWebVisitsMonth: Number of visits to the company’s website in the last month

Berikut 5 baris teratas dan terbawah dari dataset supermarket customers:

In [2]:
df = pd.read_csv('Supermarket Customers.csv', sep='\t')
display(df.head(),df.tail())

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  04-09-2012       58       635  ...                  7             0   
1  08-03-2014       38        11  ...                  5             0   
2  21-08-2013       26       426  ...                  4             0   
3  10-02-2014       26        11  ...                  6             0   
4  19-01-2014       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
2235         1  13-06-2013       46       709  ...                  5   
2236         1  10-06-2014       56       406  ...                  7   
2237         0  25-01-2014       91       908  ...                  6   
2238         1  24-01-2014        8       428  ...                  3   
2239         1  15-10-2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[5 rows x 29 columns]

## Data Understanding and Cleaning
Sebelum masuk ke dalam analisis, kita perlu mengenal dataset kita lebih jauh dalam tahapan *data understanding*. Dari proses ini, kita akan tahu anomali-anomali apa saja yang terdapat di dalam dataset kita dan perlu ditangani dalam tahapan *data cleaning*. Setiap penangan anomali yang dilakukan, akan disertai dengan justifikasi langkah yang diambil, baik secara *domain knowledge* maupun secara statistik.

Pertama, mari kita lihat informasi umum dari dataset supermarket customers.

In [3]:
print(f'Jumlah baris dan kolom di dataset df adalah {df.shape}')
df.info()

Jumlah baris dan kolom di dataset df adalah (2240, 29)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-nul

In [4]:
display(df.describe().round(3), df.describe(include='object'))

ID  Year_Birth      Income   Kidhome  Teenhome   Recency  \
count   2240.000    2240.000    2216.000  2240.000  2240.000  2240.000   
mean    5592.160    1968.806   52247.251     0.444     0.506    49.109   
std     3246.662      11.984   25173.077     0.538     0.545    28.962   
min        0.000    1893.000    1730.000     0.000     0.000     0.000   
25%     2828.250    1959.000   35303.000     0.000     0.000    24.000   
50%     5458.500    1970.000   51381.500     0.000     0.000    49.000   
75%     8427.750    1977.000   68522.000     1.000     1.000    74.000   
max    11191.000    1996.000  666666.000     2.000     2.000    99.000   

       MntWines  MntFruits  MntMeatProducts  MntFishProducts  ...  \
count  2240.000   2240.000         2240.000         2240.000  ...   
mean    303.936     26.302          166.950           37.525  ...   
std     336.597     39.773          225.715           54.629  ...   
min       0.000      0.000            0.000            0.000  ...   
25%      23.750      1.000           16.000            3.000  ...   
50%     173.500      8.000           67.000           12.000  ...   
75%     504.250     33.000          232.000           50.000  ...   
max    1493.000    199.000         1725.000          259.000  ...   

       NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
count           2240.000      2240.000      2240.000      2240.000   
mean               5.317         0.073         0.075         0.073   
std                2.427         0.260         0.263         0.260   
min                0.000         0.000         0.000         0.000   
25%                3.000         0.000         0.000         0.000   
50%                6.000         0.000         0.000         0.000   
75%                7.000         0.000         0.000         0.000   
max               20.000         1.000         1.000         1.000   

       AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  \
count      2240.000      2240.000  2240.000         2240.0     2240.0   
mean          0.064         0.013     0.009            3.0       11.0   
std           0.245         0.115     0.096            0.0        0.0   
min           0.000         0.000     0.000            3.0       11.0   
25%           0.000         0.000     0.000            3.0       11.0   
50%           0.000         0.000     0.000            3.0       11.0   
75%           0.000         0.000     0.000            3.0       11.0   
max           1.000         1.000     1.000            3.0       11.0   

       Response  
count  2240.000  
mean      0.149  
std       0.356  
min       0.000  
25%       0.000  
50%       0.000  
75%       0.000  
max       1.000  

[8 rows x 26 columns]

Education Marital_Status Dt_Customer
count         2240           2240        2240
unique           5              8         663
top     Graduation        Married  31-08-2012
freq          1127            864          12

In [5]:
# Gunakan None atau angka positif untuk max_colwidth
pd.set_option('display.max_colwidth', 1000)  # atau gunakan angka besar seperti 1000

# data unik di tiap kolom
listItem = []
for col in df.columns:
    listItem.append([col, df[col].nunique(), df[col].unique()])

tabel1Desc = pd.DataFrame(columns=['Column Name', 'Number of Unique', 'Unique Sample'],
                         data=listItem)
tabel1Desc

Column Name  Number of Unique  \
0                    ID              2240   
1            Year_Birth                59   
2             Education                 5   
3        Marital_Status                 8   
4                Income              1974   
5               Kidhome                 3   
6              Teenhome                 3   
7           Dt_Customer               663   
8               Recency               100   
9              MntWines               776   
10            MntFruits               158   
11      MntMeatProducts               558   
12      MntFishProducts               182   
13     MntSweetProducts               177   
14         MntGoldProds               213   
15    NumDealsPurchases                15   
16      NumWebPurchases                15   
17  NumCatalogPurchases                14   
18    NumStorePurchases                14   
19    NumWebVisitsMonth                16   
20         AcceptedCmp3                 2   
21         AcceptedCmp4                 2   
22         AcceptedCmp5                 2   
23         AcceptedCmp1                 2   
24         AcceptedCmp2                 2   
25             Complain                 2   
26        Z_CostContact                 1   
27            Z_Revenue                 1   
28             Response                 2   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Unique Sample  
0                                                                                                                                                                                                                                                                                                                                                                                                            [5524, 2174, 4141, 6182, 5324, 7446, 965, 6177, 4855, 5899, 1994, 387, 2125, 8180, 2569, 2114, 9736, 4939, 6565, 2278, 9360, 5376, 1993, 4047, 1409, 7892, 2404, 5255, 9422, 1966, 6864, 3033, 5710, 7373, 8755, 10738, 4339, 10755, 8595, 2968, 8601, 503, 8430, 7281, 2139, 1371, 9909, 7286, 7244, 6566, 8614, 4114, 1331, 2225, 9381, 6260, 10383, 7437, 8557, 8375, 6853, 3076, 1012, 6518, 4137, 8082, 1386, 9369, 4477, 1357, 1402, 10629, 6312, 8132, 1050, 5846, 1859, 7503, 10401, 1618, 3332, 2261, 5346, 4119, 535, 5268, 273, 4452, 8504, 771, 8996, 9235, 5798, 11178, 2230, 7516, 7247, 11100, 4646, 3037, ...]  
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

Secara umum, kita bisa melihat bahwa:
* dataset supermarket_customer memiliki 2240 baris dan 29 kolom
* Pada website [Europian union](https://ec.europa.eu/eurostat/statistics-explained/index.php?title=Glossary:Marital_status#:~:text=There%20are%20several%20types%20of,in%20certain%20cases%2C%20registered%20partnership.) menunjukkan bahwa pada kolom `Marital_Status`: Status pernikahan pelanggan. Beberapa nilai seperti YOLO dan Absurd tidak relevan dan akan dikategorikan sebagai NaN. Beberapa kategori seperti Together bisa digabungkan dengan Married, dan Alone dengan Single.
* Beberapa kolom seperti `Income`, `Recency`, `Year_Birth`. Perlu dicek kembali agar tidak ada *outlier*
* Pada kolom `Income`, ada *missing value*.
* Adapun beberapa kolom yang dapat diabaikan seperti `Dt_Customer`, `Z_CostContact`, `Z_Revenue`, `ID` karena tidak memiliki pengaruh terhadap analisis ini.
* Pada kolom `Education` akan digabungkan dengan Master karena 2n Cycle memiliki nilai yang sama dengan Master.
* Pada `MntWines`, `MntFruits`, `MntMeatProducts`, `MntFishProducts`, `MntSweetProducts`, `MntGoldProds`. Jumlah uang yang dihabiskan pelanggan untuk kategori produk tertentu dalam dua tahun terakhir. Perlu dicek apakah ada nilai ekstrem atau anomali.
* `NumDealsPurchases` Bisa digunakan untuk melihat apakah pelanggan cenderung membeli hanya saat ada promo.
* `Complain` Bisa digunakan untuk melihat hubungan dengan kepuasan pelanggan.
* Diketahui bahwa terakhir pelanggan berbelanja adalah pada tanggal 2014-09-08. maka dapat diasumsikan bahwa data terakhir didapat pada tahun 2014. sehingga kita dapat menambahkan kolom baru yaitu `Year_old` untuk mengetahui umur pelanggan saat ini, dengan cara `Year_old = 2014 - Year_Birth`.

### Missing Value

In [6]:
# persentase missing value di tiap kolom
df.isna().sum()/df.shape[0]*100

ID                     0.000000
Year_Birth             0.000000
Education              0.000000
Marital_Status         0.000000
Income                 1.071429
Kidhome                0.000000
Teenhome               0.000000
Dt_Customer            0.000000
Recency                0.000000
MntWines               0.000000
MntFruits              0.000000
MntMeatProducts        0.000000
MntFishProducts        0.000000
MntSweetProducts       0.000000
MntGoldProds           0.000000
NumDealsPurchases      0.000000
NumWebPurchases        0.000000
NumCatalogPurchases    0.000000
NumStorePurchases      0.000000
NumWebVisitsMonth      0.000000
AcceptedCmp3           0.000000
AcceptedCmp4           0.000000
AcceptedCmp5           0.000000
AcceptedCmp1           0.000000
AcceptedCmp2           0.000000
Complain               0.000000
Z_CostContact          0.000000
Z_Revenue              0.000000
Response               0.000000
dtype: float64

Dikarenakan *missing value* pada kolom hanya ada pada kolom `income` dengan proporsi yang sangat kecil (<2%), kita bisa menghapus baris yang mengandung *missing value* tersebut.

In [7]:
# Menghapus Missing value
df.dropna(inplace=True)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2216 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2216 non-null   int64  
 1   Year_Birth           2216 non-null   int64  
 2   Education            2216 non-null   object 
 3   Marital_Status       2216 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2216 non-null   int64  
 6   Teenhome             2216 non-null   int64  
 7   Dt_Customer          2216 non-null   object 
 8   Recency              2216 non-null   int64  
 9   MntWines             2216 non-null   int64  
 10  MntFruits            2216 non-null   int64  
 11  MntMeatProducts      2216 non-null   int64  
 12  MntFishProducts      2216 non-null   int64  
 13  MntSweetProducts     2216 non-null   int64  
 14  MntGoldProds         2216 non-null   int64  
 15  NumDealsPurchases    2216 non-null   int64 

Setelah semua jumlah data disamakan maka dapat dilakukan eksplorasi data lebih lanjut.

### Menghapus Kolom
Pertama, kita akan menghapus kolom yang tidak dibutuhkan untuk analisis ini. Kolom yang akan dihapus adalah: `ID`. Kolom ini tidak memiliki informasi yang relevan untuk analisis ini.

Untuk kolom `Dt_Customer`, `Z_CostContact`, dan `Z_Revenue` juga akan kenapa dihapus adalah karena tujuan analisis ini adalah untuk mengetahui pola pembelian pelanggan. Oleh karena itu, membandingkan variabel yang tidak secara langsung berkaitan dengan kebiasaan belanja pelanggan tidak akan memberikan wawasan yang relevan.

In [9]:
df.drop(['ID', 'Dt_Customer', 'Z_CostContact', 'Z_Revenue'], axis=1, inplace=True)

In [10]:
df.columns

Index(['Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Recency', 'MntWines', 'MntFruits', 'MntMeatProducts',
       'MntFishProducts', 'MntSweetProducts', 'MntGoldProds',
       'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases',
       'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3',
       'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2',
       'Complain', 'Response'],
      dtype='object')

### Memperbaiki kolom
Pada kolom `Marital_Status` kita harus menangani nilai yang tidak valid dengan cara mengubah data 'Alone' menjadi 'Single' dan 'Together' menjadi 'Married'. Kemudian menghapus data 'Absurd' dan 'YOLO' menjadi NaN.

In [11]:
df['Marital_Status'].value_counts()

Marital_Status
Married     857
Together    573
Single      471
Divorced    232
Widow        76
Alone         3
Absurd        2
YOLO          2
Name: count, dtype: int64

In [12]:
# Menggabungkan "Alone" dan "Together" menjadi "Single" dan "Married"
df.loc[df['Marital_Status']=='Alone', 'Marital_Status'] = 'Single'
df.loc[df['Marital_Status']=='Together', 'Marital_Status'] = 'Married'

# Mengganti "YOLO" dan "Absurd" dengan NaN
df['Marital_Status'] = df['Marital_Status'].replace(['YOLO', 'Absurd'], 'Unknown')

In [13]:
df['Marital_Status'].value_counts()

Marital_Status
Married     1430
Single       474
Divorced     232
Widow         76
Unknown        4
Name: count, dtype: int64

### Menambahkan kolom
Disini saya menambahkan kolom baru yakni:

* `Age` : Dapat dihitung dari `Year_Birth`, untuk mengetahui usia pelanggan.
* `Total_spending` : Dapat dihitung dari `MntWines`, `MntFruits`, `MntMeatProducts`, `MntFishProducts`, `MntSweetProducts`, `MntGoldProds` untuk mengetahui total pengeluaran pelanggan.
* `Avg_spending` : Dapat dihitung dari `Total_spending` dan `NumWebPurchases` untuk mengetahui rata-rata pengeluaran pelanggan.
* `Family_Size` : Dapat dihitung dari `Kidhome` dan `Teenhome` untuk mengetahui ukuran keluarga pelanggan.
* `Total_Accepted_campaigns` : Dapat dihitung dari `AcceptedCmp1`, `AcceptedCmp2`, `AcceptedCmp3`, `AcceptedCmp4`, `AcceptedCmp5`, `Response` untuk mengetahui total kampanye yang diikuti pelanggan.

Age: 

Bisa digunakan untuk melihat apakah pelanggan yang lebih tua atau lebih muda memiliki kebiasaan belanja yang berbeda.

In [14]:
df['Age'] = 2014 - df['Year_Birth']

Total_Spending:

Bisa digunakan untuk mengelompokkan pelanggan berdasarkan tingkat pengeluaran (low spender vs high spender).

In [15]:
df['Total_Spending'] = df[['MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds']].sum(axis=1)


Avg_Spending:

Bisa membantu melihat apakah pelanggan dengan pengeluaran tinggi melakukan sedikit pembelian besar atau banyak pembelian kecil.

In [16]:
df['Total_Purchases'] = df[['NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases']].sum(axis=1)
df['Avg_Spending_Per_Purchase'] = (df['Total_Spending'] / df['Total_Purchases']).round(2)

Family_Size:

Bisa digunakan untuk melihat apakah keluarga besar memiliki kebiasaan belanja berbeda dari pelanggan yang tinggal sendiri

In [17]:
df['Family_Size'] = df['Kidhome'] + df['Teenhome'] + 1

Total_Accepted_Campaigns:

Bisa membantu melihat apakah pelanggan yang sering menerima promo memiliki kebiasaan belanja yang berbeda.

In [18]:
df['Total_Accepted_Campaigns'] = df[['AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'Response']].sum(axis=1)

## Data yang sudah bersih
Kita sudah selesai membersihkan data kita, sekarang, mari kita lihat sekilas informasi dari data yang sudah kita bersihkan

Mengecek kembali isi keseluruhan data:

In [19]:
display(df.head(), df.tail())

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  Recency  \
0        1957  Graduation         Single  58138.0        0         0       58   
1        1954  Graduation         Single  46344.0        1         1       38   
2        1965  Graduation        Married  71613.0        0         0       26   
3        1984  Graduation        Married  26646.0        1         0       26   
4        1981         PhD        Married  58293.0        1         0       94   

   MntWines  MntFruits  MntMeatProducts  ...  AcceptedCmp1  AcceptedCmp2  \
0       635         88              546  ...             0             0   
1        11          1                6  ...             0             0   
2       426         49              127  ...             0             0   
3        11          4               20  ...             0             0   
4       173         43              118  ...             0             0   

   Complain  Response  Age  Total_Spending  Total_Purchases  \
0         0         1   57            1617               22   
1         0         0   60              27                4   
2         0         0   49             776               20   
3         0         0   30              53                6   
4         0         0   33             422               14   

   Avg_Spending_Per_Purchase  Family_Size  Total_Accepted_Campaigns  
0                      73.50            1                         1  
1                       6.75            3                         0  
2                      38.80            1                         0  
3                       8.83            2                         0  
4                      30.14            2                         0  

[5 rows x 31 columns]

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
2235        1967  Graduation        Married  61223.0        0         1   
2236        1946         PhD        Married  64014.0        2         1   
2237        1981  Graduation       Divorced  56981.0        0         0   
2238        1956      Master        Married  69245.0        0         1   
2239        1954         PhD        Married  52869.0        1         1   

      Recency  MntWines  MntFruits  MntMeatProducts  ...  AcceptedCmp1  \
2235       46       709         43              182  ...             0   
2236       56       406          0               30  ...             1   
2237       91       908         48              217  ...             0   
2238        8       428         30              214  ...             0   
2239       40        84          3               61  ...             0   

      AcceptedCmp2  Complain  Response  Age  Total_Spending  Total_Purchases  \
2235             0         0         0   47            1341               16   
2236             0         0         0   68             444               15   
2237             0         0         0   33            1241               18   
2238             0         0         0   58             843               21   
2239             0         0         1   60             172                8   

      Avg_Spending_Per_Purchase  Family_Size  Total_Accepted_Campaigns  
2235                      83.81            2                         0  
2236                      29.60            4                         1  
2237                      68.94            1                         1  
2238                      40.14            2                         0  
2239                      21.50            3                         1  

[5 rows x 31 columns]

In [20]:
listItem = []
for col in df.columns :
    listItem.append([col, df[col].dtype, df[col].isna().sum(),
                    df[col].nunique(), list(df[col].drop_duplicates().sample(2).values)]);

dfDesc = pd.DataFrame(columns=['dataFeatures', 'dataType', 'null', 'unique', 'uniqueSample'],
                     data=listItem)
print(df.shape)
dfDesc

(2216, 31)


dataFeatures dataType  null  unique        uniqueSample
0                  Year_Birth    int64     0      59        [1961, 1984]
1                   Education   object     0       5     [Master, Basic]
2              Marital_Status   object     0       5   [Married, Single]
3                      Income  float64     0    1974  [38725.0, 34853.0]
4                     Kidhome    int64     0       3              [1, 0]
5                    Teenhome    int64     0       3              [2, 1]
6                     Recency    int64     0     100            [85, 87]
7                    MntWines    int64     0     776          [106, 897]
8                   MntFruits    int64     0     158           [81, 197]
9             MntMeatProducts    int64     0     554          [227, 264]
10            MntFishProducts    int64     0     182           [41, 210]
11           MntSweetProducts    int64     0     176            [11, 41]
12               MntGoldProds    int64     0     212           [35, 129]
13          NumDealsPurchases    int64     0      15             [4, 10]
14            NumWebPurchases    int64     0      15             [1, 11]
15        NumCatalogPurchases    int64     0      14              [4, 5]
16          NumStorePurchases    int64     0      14              [5, 6]
17          NumWebVisitsMonth    int64     0      16             [20, 8]
18               AcceptedCmp3    int64     0       2              [1, 0]
19               AcceptedCmp4    int64     0       2              [1, 0]
20               AcceptedCmp5    int64     0       2              [1, 0]
21               AcceptedCmp1    int64     0       2              [1, 0]
22               AcceptedCmp2    int64     0       2              [0, 1]
23                   Complain    int64     0       2              [1, 0]
24                   Response    int64     0       2              [1, 0]
25                        Age    int64     0      59            [56, 28]
26             Total_Spending    int64     0    1047          [614, 642]
27            Total_Purchases    int64     0      33            [20, 23]
28  Avg_Spending_Per_Purchase  float64     0    1257    [114.07, 127.29]
29                Family_Size    int64     0       4              [3, 4]
30   Total_Accepted_Campaigns    int64     0       6              [4, 0]

# Data Analysis
Kita sudah melakukan tahap _data cleaning_. Sekarang, kita bisa mulai melakukan analisis untuk mencari tahu **Bagaimana karakteristik pelanggan yang memberikan kontribusi terbesar terhadap pendapatan supermarket, terutama dibandingkan pelanggan dengan pengeluaran lebih rendah?**.

Analisis akan dilakukan dengan membandingkan pelanggan berdasarkan pola pembelian dan karakteristiknya. Tujuan utama adalah memahami faktor-faktor yang memengaruhi perilaku belanja pelanggan.

Dari **2216 data pelanggan**, kita akan melihat perbedaan perilaku berdasarkan berbagai fitur seperti **umur, status pernikahan, tingkat pendidikan, pendapatan, serta kebiasaan belanja**.

Karena data ini memiliki berbagai fitur numerik dan kategorikal, maka:

* **Untuk data kategorikal**, perbandingan akan dilakukan berdasarkan **proporsi (%)**, bukan hanya jumlah absolut, agar lebih akurat dalam memahami distribusi.
* **Untuk data numerik**, analisis dilakukan dengan melihat **rata-rata, median, serta distribusi (skewness & outliers)** guna memahami kecenderungan dan variasi antar pelanggan.
Selain itu, kita juga akan melihat *p*ola pembelian dan efektivitas kampanye pemasaran**, untuk mengetahui apakah ada perbedaan signifikan antara pelanggan yang menerima promo dengan yang tidak.

## **4.1 Question 1**

### **4.2 Question 2**

## **Section 5. Conclusion and Recommendation**

**5.1 Conclusion**

**5.2 Recommendation**